include/sparrow/buffer_view.hpp

```diff
@@ -0,0 +1,270 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "sparrow/buffer.hpp"
+
+namespace sparrow
+{
+    /*
+     * @class buffer_view
+     * @brief Object that references but does not own a piece of contiguous memory
+     */
+    template <class T>
+    class buffer_view
+    {
+    public:
+
+        using self_type = buffer_view<T>;
+        using value_type = T;
+        using reference = T&;
+        using const_reference = const T&;
+        using pointer = T*;
+        using const_pointer = const T*;
+        using size_type = std::size_t;
+        using difference_type = std::ptrdiff_t;
+
+        using iterator = pointer_iterator<pointer>;
+        using const_iterator = pointer_iterator<const_pointer>;
+        using reverse_iterator = std::reverse_iterator<iterator>;
+        using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+        explicit buffer_view(buffer<T>& buffer);
+        buffer_view(pointer p, size_type n);
```

Comment:
To be complete we also need a `subrange(...)` function and a constructor that takes a pair of iterators (they should be template parameters constrained by concept, not `buffer_view::iterator`).
These are often used with view types (see `std::span`, which also provides `first` and `last` subview functions).

Reply:
Maybe we can do that in a dedicated PR? This piece of code was just moved from the initial `buffer.hpp` file. I'll add a todo for that.

Reply:
I feel that this and the iterator pair constructor (which should be what the subrange is built with) not being there makes it incomplete. But ok, it can be added in another PR although please do it immediately after this one then?